diff --git a/org/apache/felix/bundlerepository/impl/LocalRepositoryImpl.java b/org/apache/felix/bundlerepository/impl/LocalRepositoryImpl.java
index 5d52f3a..17e4228 100644
--- a/org/apache/felix/bundlerepository/impl/LocalRepositoryImpl.java
+++ b/org/apache/felix/bundlerepository/impl/LocalRepositoryImpl.java
@@ -47,7 +47,9 @@ public class LocalRepositoryImpl implements Repository, SynchronousBundleListene
 
     public void bundleChanged(BundleEvent event)
     {
-        if (event.getType() == BundleEvent.INSTALLED)
+		Bundle bundle = event.getBundle();
+
+        if ((event.getType() == BundleEvent.INSTALLED) && (bundle.getState() != Bundle.UNINSTALLED))
         {
             synchronized (this)
             {
