diff --git a/org/apache/felix/fileinstall/internal/DirectoryWatcher.java b/org/apache/felix/fileinstall/internal/DirectoryWatcher.java
index cfbdab1cc6d5..c0baa891844b 100644
--- a/org/apache/felix/fileinstall/internal/DirectoryWatcher.java
+++ b/org/apache/felix/fileinstall/internal/DirectoryWatcher.java
@@ -1224,10 +1224,10 @@ public class DirectoryWatcher extends Thread implements BundleListener
     private void startBundles(Collection<Bundle> bundles)
     {
         // Check if this is the consistent set of bundles which failed previously.
-        boolean logFailures = bundles.equals(consistentlyFailingBundles);
+
         for (Iterator<Bundle> b = bundles.iterator(); b.hasNext(); )
         {
-            if (startBundle(b.next(), logFailures))
+            if (startBundle(b.next(), true))
             {
                 b.remove();
             }
@@ -1264,7 +1264,7 @@ public class DirectoryWatcher extends Thread implements BundleListener
                 // Don't log this as an error, instead we start the bundle repeatedly.
                 if (logFailures)
                 {
-                    log(Logger.LOG_WARNING, "Error while starting bundle: " + bundle.getLocation(), e);
+                    log(Logger.LOG_ERROR, "Error while starting bundle: " + bundle.getLocation(), e);
                 }
             }
         }
