diff --git a/org/apache/felix/fileinstall/internal/DirectoryWatcher.java b/org/apache/felix/fileinstall/internal/DirectoryWatcher.java
index 629e48dca6ea..ebe7251ed084 100644
--- a/org/apache/felix/fileinstall/internal/DirectoryWatcher.java
+++ b/org/apache/felix/fileinstall/internal/DirectoryWatcher.java
@@ -1035,6 +1035,11 @@ public class DirectoryWatcher extends Thread implements BundleListener
             Util.log(context, Logger.LOG_INFO, "Installing bundle " + sn
                     + " / " + v, null);
             Bundle b = context.installBundle(bundleLocation, is);
+
+            if (b.getState() == Bundle.UNINSTALLED) {
+                return b;
+            }
+
             Util.storeChecksum(b, checksum, context);
             modified.set(true);
 
@@ -1496,3 +1501,4 @@ public class DirectoryWatcher extends Thread implements BundleListener
     }
 
 }
+/* @generated */
