diff --git a/org/apache/felix/scr/impl/manager/DependencyManager.java b/org/apache/felix/scr/impl/manager/DependencyManager.java
index 7076601..a44e28b 100644
--- a/org/apache/felix/scr/impl/manager/DependencyManager.java
+++ b/org/apache/felix/scr/impl/manager/DependencyManager.java
@@ -1577,7 +1577,9 @@ public class DependencyManager<S, T> implements ReferenceManager<S, T>
     {
         // only invoke the unbind method if there is an instance (might be null
         // in the delayed component situation) and the unbind method is declared.
-        boolean doUnbind = componentContext != null && m_dependencyMetadata.getUnbind() != null;
+        boolean doUnbind =
+        	componentContext != null &&
+        		(m_dependencyMetadata.getField() != null || m_dependencyMetadata.getUnbind() != null);

         AtomicInteger trackingCount = new AtomicInteger();
         Collection<RefPair<S, T>> refPairs;
@@ -2226,3 +2228,4 @@ public class DependencyManager<S, T> implements ReferenceManager<S, T>
     }

 }
+/* @generated */
