diff --git a/org/apache/felix/cm/impl/ConfigurationImpl.java b/org/apache/felix/cm/impl/ConfigurationImpl.java
index 4d97ce7..a6c695c 100644
--- a/org/apache/felix/cm/impl/ConfigurationImpl.java
+++ b/org/apache/felix/cm/impl/ConfigurationImpl.java
@@ -463,6 +463,8 @@ public class ConfigurationImpl
         // finally assign the configuration for use
         configure( newProperties );
 
+        configurationManager.removeConfiguration(this);
+
         // update the service and fire an CM_UPDATED event
         configurationManager.updated( this, true );
     }
@@ -908,3 +910,4 @@ public class ConfigurationImpl
         return this.configurationManager;
     }
 }
+/* @generated */
\ No newline at end of file
