diff --git a/org/hibernate/cfg/Configuration.java b/org/hibernate/cfg/Configuration.java
index eb369b0..d9fc395 100644
--- a/org/hibernate/cfg/Configuration.java
+++ b/org/hibernate/cfg/Configuration.java
@@ -1858,10 +1858,6 @@ public class Configuration implements Serializable {
 			log.warn( "mapping metadata cache was not completely processed" );
 		}
 
-		enableLegacyHibernateValidator();
-		enableBeanValidation();
-		enableHibernateSearch();
-
 		validate();
 		Environment.verifyProperties( properties );
 		Properties copy = new Properties();
@@ -4146,3 +4142,4 @@ public class Configuration implements Serializable {
 		public boolean cacheLazy;
 	}
 }
+/* @generated */
\ No newline at end of file
