diff --git a/org/hibernate/cfg/SettingsFactory.java b/org/hibernate/cfg/SettingsFactory.java
index ae5d22e..1ef8619 100644
--- a/org/hibernate/cfg/SettingsFactory.java
+++ b/org/hibernate/cfg/SettingsFactory.java
@@ -122,19 +122,6 @@ public class SettingsFactory implements Serializable {
 					metaReportsDDLCausesTxnCommit = meta.dataDefinitionCausesTransactionCommit();
 					metaReportsDDLInTxnSupported = !meta.dataDefinitionIgnoredInTransactions();
 					metaSupportsGetGeneratedKeys = meta.supportsGetGeneratedKeys();
-					
-					log.info( "Database ->\n" +
-							"       name : " + meta.getDatabaseProductName() + '\n' +
-							"    version : " +  meta.getDatabaseProductVersion() + '\n' +
-							"      major : " + meta.getDatabaseMajorVersion() + '\n' +
-							"      minor : " + meta.getDatabaseMinorVersion()
-					);
-					log.info( "Driver ->\n" +
-							"       name : " + meta.getDriverName() + '\n' +
-							"    version : " + meta.getDriverVersion() + '\n' +
-							"      major : " + meta.getDriverMajorVersion() + '\n' +
-							"      minor : " + meta.getDriverMinorVersion()
-					);
 				}
 				catch ( SQLException sqle ) {
 					log.warn( "Could not obtain connection metadata", sqle );
@@ -468,3 +455,4 @@ public class SettingsFactory implements Serializable {
 	}
 
 }
+/* @generated */
\ No newline at end of file
