diff --git a/org/hibernate/engine/jdbc/BlobProxy.java b/org/hibernate/engine/jdbc/BlobProxy.java
index f81e4b4..16e1d58 100644
--- a/org/hibernate/engine/jdbc/BlobProxy.java
+++ b/org/hibernate/engine/jdbc/BlobProxy.java
@@ -192,10 +192,6 @@ public class BlobProxy implements InvocationHandler {
 	 * @return The class loader appropriate for proxy construction.
 	 */
 	private static ClassLoader getProxyClassLoader() {
-		ClassLoader cl = Thread.currentThread().getContextClassLoader();
-		if ( cl == null ) {
-			cl = BlobImplementer.class.getClassLoader();
-		}
-		return cl;
+		return BlobImplementer.class.getClassLoader();
 	}
 }
+/* @generated */
\ No newline at end of file
diff --git a/org/hibernate/engine/jdbc/ClobProxy.java b/org/hibernate/engine/jdbc/ClobProxy.java
index 97ae85a..0302dfd 100644
--- a/org/hibernate/engine/jdbc/ClobProxy.java
+++ b/org/hibernate/engine/jdbc/ClobProxy.java
@@ -218,10 +218,6 @@ public class ClobProxy implements InvocationHandler {
 	 * @return The class loader appropriate for proxy construction.
 	 */
 	protected static ClassLoader getProxyClassLoader() {
-		ClassLoader cl = Thread.currentThread().getContextClassLoader();
-		if ( cl == null ) {
-			cl = ClobImplementer.class.getClassLoader();
-		}
-		return cl;
+		return ClobImplementer.class.getClassLoader();
 	}
 }
+/* @generated */
\ No newline at end of file
diff --git a/org/hibernate/engine/jdbc/NClobProxy.java b/org/hibernate/engine/jdbc/NClobProxy.java
index e5a1739..a624b02 100644
--- a/org/hibernate/engine/jdbc/NClobProxy.java
+++ b/org/hibernate/engine/jdbc/NClobProxy.java
@@ -96,10 +96,6 @@ public class NClobProxy extends ClobProxy {
 	 * @return The class loader appropriate for proxy construction.
 	 */
 	protected static ClassLoader getProxyClassLoader() {
-		ClassLoader cl = Thread.currentThread().getContextClassLoader();
-		if ( cl == null ) {
-			cl = NClobImplementer.class.getClassLoader();
-		}
-		return cl;
+		return NClobImplementer.class.getClassLoader();
 	}
 }
+/* @generated */
\ No newline at end of file
diff --git a/org/hibernate/engine/jdbc/SerializableBlobProxy.java b/org/hibernate/engine/jdbc/SerializableBlobProxy.java
index 1f857ee..2c4dfc8 100644
--- a/org/hibernate/engine/jdbc/SerializableBlobProxy.java
+++ b/org/hibernate/engine/jdbc/SerializableBlobProxy.java
@@ -103,10 +103,6 @@ public class SerializableBlobProxy implements InvocationHandler, Serializable {
 	 * @return The class loader appropriate for proxy construction.
 	 */
 	public static ClassLoader getProxyClassLoader() {
-		ClassLoader cl = Thread.currentThread().getContextClassLoader();
-		if ( cl == null ) {
-			cl = WrappedBlob.class.getClassLoader();
-		}
-		return cl;
+		return WrappedBlob.class.getClassLoader();
 	}
 }
+/* @generated */
\ No newline at end of file
diff --git a/org/hibernate/engine/jdbc/SerializableClobProxy.java b/org/hibernate/engine/jdbc/SerializableClobProxy.java
index 87ee1ee..7e63404 100644
--- a/org/hibernate/engine/jdbc/SerializableClobProxy.java
+++ b/org/hibernate/engine/jdbc/SerializableClobProxy.java
@@ -102,10 +102,6 @@ public class SerializableClobProxy implements InvocationHandler, Serializable {
 	 * @return The class loader appropriate for proxy construction.
 	 */
 	public static ClassLoader getProxyClassLoader() {
-		ClassLoader cl = Thread.currentThread().getContextClassLoader();
-		if ( cl == null ) {
-			cl = WrappedClob.class.getClassLoader();
-		}
-		return cl;
+		return WrappedClob.class.getClassLoader();
 	}
 }
+/* @generated */
\ No newline at end of file
