diff --git a/org/apache/felix/gogo/runtime/threadio/ThreadIOImpl.java b/org/apache/felix/gogo/runtime/threadio/ThreadIOImpl.java
index 340a533..27f32f5 100644
--- a/org/apache/felix/gogo/runtime/threadio/ThreadIOImpl.java
+++ b/org/apache/felix/gogo/runtime/threadio/ThreadIOImpl.java
@@ -87,9 +87,9 @@ public class ThreadIOImpl implements ThreadIO
 		if (markers.isEmpty()) {
 			current.remove();
 
-			System.setOut(defaultMarker.out);
-			System.setIn(defaultMarker.in);
-			System.setErr(defaultMarker.err);
+			System.setOut(this.out);
+			System.setIn(this.in);
+			System.setErr(this.err);
 		}
     }
 
