diff --git a/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java b/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java
index 6caa4e4a6608..abff18f53688 100644
--- a/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java
+++ b/org/elasticsearch/painless/lookup/PainlessLookupBuilder.java
@@ -54,7 +54,6 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.regex.Pattern;
 
 import static org.elasticsearch.painless.WriterConstants.DEF_TO_B_BYTE_IMPLICIT;
 import static org.elasticsearch.painless.WriterConstants.DEF_TO_B_CHARACTER_IMPLICIT;
@@ -1106,7 +1105,7 @@ public final class PainlessLookupBuilder {
                     "cannot represent multiple java classes with the same name from different class loaders");
         }
 
-        if (METHOD_NAME_PATTERN.matcher(methodName).matches() == false) {
+        if (!_isValidName(methodName, false)) {
             throw new IllegalArgumentException(
                     "invalid method name [" + methodName + "] for instance binding [" + targetCanonicalClassName + "].");
         }
