diff --git a/org/apache/jasper/compiler/SmapUtil.java b/org/apache/jasper/compiler/SmapUtil.java
index 00fec746eb77..4e8a4e2c2221 100644
--- a/org/apache/jasper/compiler/SmapUtil.java
+++ b/org/apache/jasper/compiler/SmapUtil.java
@@ -430,7 +430,11 @@ public class SmapUtil {
                 switch (tag) {
                     case 7 : // Class
                     case 8 : // String
+					case 16 : // MethodType
                         copy(2);
+                        break;
+					case 15 : // MethodHandle
+                        copy(3);
                         break;
                     case 9 : // Field
                     case 10 : // Method
@@ -438,6 +442,7 @@ public class SmapUtil {
                     case 3 : // Integer
                     case 4 : // Float
                     case 12 : // NameAndType
+					case 18 : // InvokeDynamic
                         copy(4);
                         break;
                     case 5 : // Long
@@ -741,3 +746,4 @@ public class SmapUtil {
     }
 
 }
+/* @generated */
\ No newline at end of file
