diff --git a/freemarker/ext/jsp/JspWriterAdapter.java b/freemarker/ext/jsp/JspWriterAdapter.java
index ad53eed..2bab201 100644
--- a/freemarker/ext/jsp/JspWriterAdapter.java
+++ b/freemarker/ext/jsp/JspWriterAdapter.java
@@ -24,7 +24,7 @@ import javax.servlet.jsp.JspWriter;
 import freemarker.template.utility.SecurityUtilities;
 
 class JspWriterAdapter extends JspWriter {
-    static final char[] NEWLINE = SecurityUtilities.getSystemProperty("line.separator", "\n").toCharArray();
+    static final String NEWLINE = SecurityUtilities.getSystemProperty("line.separator", "\n");
     
     private final Writer out;
     
@@ -174,4 +174,16 @@ class JspWriterAdapter extends JspWriter {
     {
         out.write(arg0, arg1, arg2);
     }
-}
\ No newline at end of file
+
+    @Override
+    public void write(String s) throws IOException {
+        out.write(s);
+    }
+
+    @Override
+    public void write(String s, int offset, int length) throws IOException {
+        out.write(s, offset, length);
+    }
+
+}
+/* @generated */
\ No newline at end of file
